File: Algorithm/Code_Materials/c_or_cpp/charpter1_linklist/level2/topic2_4双指针/middle_node.cpp
#include <stdlib.h>
#include <stdio.h>

#include <stdio.h>
#include "../../list_node.cpp"

struct ListNode *middleNode(struct ListNode *head)
{
    struct ListNode *slow = head;
    struct ListNode *fast = head;

    while (fast != NULL && fast->next != NULL)
    {
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow;
}

int main()
{
    struct ListNode *list1 = NULL;

    int a[] = {1, 2, 3, 4, 5};

    list1 = initLink(a, 5);

    struct ListNode *list3 = middleNode(list1);

    // 输出的是3 4 5,我们主要看第一个是3就可以了，不必纠结后面的结果，想一下为什么
    printList(list3);
    return 0;
}

File: Algorithm/Leetcode/Code_Materials/python_source/charpter1_linklist/level2/topic2_4双指针/middle_node.py
# coding=utf-8

from python_source.list_node import init_list

class MiddleNode:
    def middleNode(self, head):
        if head is None:
            return None

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow


if __name__ == '__main__':
    # 先构造题目要求的链表,简单构造即可
    nums = [1, 2, 3, 4, 5, 6]
    list = init_list(nums)
    middleNode = MiddleNode()
    node = middleNode.middleNode(list)
    print node.val
